### DISCUSSION TITLE
How to Mock a Prisma Client?

### DISCUSSION THREAD

**Discussion by 0916dhkim** [prisma/prisma#4655] [Q&A] [answered]
Integration testing of a GraphQL server usually requires a dev database. I have read documentation of Nexus & Apollo server, and both requires a connection to an actual database.
 - [Nexus.js](https://nexusjs.org/docs/getting-started/tutorial/chapter-testing-with-prisma)
 - [Apollo server](https://www.apollographql.com/docs/apollo-server/testing/testing/)

Is there a way to mock prisma client or run tests without spinning up a database?

I attempted to replace all properties of `PrismaClient.prototype` with jest mock functions:
```ts
const mockPrisma = Object.create(PrismaClient.prototype);
for (const k in mockPrisma) {
  mockPrisma[k] = jest.fn();
}
```
The code above does not work because the constructor of `PrismaClient` dynamically adds model properties (so prototype does not have model property names defined).
Here's an PR in which includes my attempt to mock `PrismaClient`. It also contains a test file that demonstrates how I want to use mocked `PrismaClient`.

**Comment by ohimors** [comment]
Any traction on this?

**Reply by 0916dhkim** [reply]
@ohimors I'm still waiting for an answer. For now, I connect to a real database for testing locally.

**Comment by nikolasburk** [comment, marked as answer]
Here's an in-depth [guide](https://dev.to/eddeee888/how-to-write-tests-for-prisma-with-docker-and-jest-593i) that covers best practices for testing with Prisma. Let me know if that helps or if you have any further questions 🙌

**Reply by 0916dhkim** [reply]
Using Docker Compose + GIthub Actions seems like a great solution to me. Thanks for sharing!

**Reply by danielzzz** [reply]
have you found any option to mock '@prisma/client' so you don't have to inject the actual prisma object into each functions?

I was trying with: 

```
jest.mock('@prisma/client', ()=>{
  return {
    PrismaClient: function() {
      return {
        performance: {
          findFirst: jest.fn(),
          update: jest.fn(),
        },
      };
    },
  };
});
```

but having trouble actually adressing the internal methods in tests.

**Reply by Jolg42** [reply]
@danielzzz You could try this https://github.com/prisma/prisma/discussions/2083

**Reply by jasonlimantoro** [reply]
I use your exact same method @danielzzz and experienced the same issues regarding the internal methods.
Finally, I got it around by declaring the module factory **above** all import statements.
Strictly speaking, it need not to be above all import statements, it just needs to be above the import statement that will eventually import the `prisma` client.

https://stackoverflow.com/a/67114668/15643119

**Reply by danielzzz** [reply]
cool, thanks for sharing!

**Reply by missbruni** [reply]
if you just want to mock without asserting the methods, instead of adding your table name and functions you can just use deepMock and everything will be mocked for you based on your schema.
```ts
import { mockDeep } from 'jest-mock-extended';
import { PrismaClient as OriginalPrismaClient } from '@prisma/client';

jest.mock('@prisma/client', () => ({
  PrismaClient: function () {
    return mockDeep<OriginalPrismaClient>();
  }
}));
```

**Reply by jorgedonoso** [reply]
> but having trouble actually adressing the internal methods in tests.

@danielzzz this is a simple way to target your Prisma methods (if you don't want to mockDeep):

```javascript
// Before test.
jest.spyOn(prisma.performance, "update").mockImplementation();
```

```javascript
// In test.
expect(prisma.performance.update).toHaveBeenCalledTimes(1);
```

**Reply by reeshabhranjan** [reply]
@jorgedonoso When I am trying to mock the `findUniqueOrThrow` method, I get this:

```Cannot spy the findUniqueOrThrow property because it is not a function; undefined given instead```

**Comment by terrierscript** [comment]
FYI: I can generate mock PrismaClient with middleware.
* https://www.prisma.io/docs/concepts/components/prisma-client/middleware

```ts
// task.ts
import { PrismaClient } from '@prisma/client'

export const getTaskById = (prisma: PrismaClient, id: string) => {
  return prisma.task.findUnique({ where: { id } })
}
```

```ts
// task.test.ts
const mockedPrismaClient = () => {
  const prismaClient = new PrismaClient()
  prismaClient.$use(async (params, next) => {
    return {
      id: "foo",
      name: "hoge"
    } as Task
  })
  return prismaClient
}

test("Task test", async () => {
  const prismaClient = mockedPrismaClient()
  const task = await getTaskById(prismaClient, "foo")
  expect(task?.name).toBe("hoge")
})
```

**Reply by 0916dhkim** [reply]
Oh wow, that's a clever way to extend an existing client :+1:

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
